### ISSUE TITLE
Why the multichannel data is randomly processed by the " chime4/asr1_multich " in v0.4.0

### ISSUE THREAD

**Issue by Rpersie** [espnet/espnet#764]
I check the procedures in the  chime4/asr1_multich of v0.4.0 and have some questions about the code in line 90 of espnet/espnet/nets/pytorch_backend/frontends/frontend.py.
![image](https://user-images.githubusercontent.com/12673020/58379540-73435700-7fd7-11e9-9bcb-062a2760244f.png)

It means that in the training "use_beamformer"  is randomly set true or false. So not all the data is processed by  the beamformer.   
I just change the code and find the loss is worse than the original code. Can anyone tell me why the  data is processed like this. In my opinion, the beamformer is not trained by all the data. The results can not perform better than using all the data 

Thank you

**Comment by kamo-naoyuki** [comment]
It have made better result experimentally, just following previous research. 

It is said that if always the beamformer part is trained, the backend asr system cannot be trained with noisy data, without any enhancement, then it  weakens the base recognition performance for noisy data.

And, there is no need to care about not using all data, in an epoch, because almost all data are given over all epoch.

**Comment by Rpersie** [comment]
How about the gap? It will behind the custom delay-sum beamformer ?  You do this just to improve the ASR performance just like the MCT.

**Comment by kamo-naoyuki** [comment]
I forgot the precise value, maybe 3~5% worse in CER.

**Comment by Rpersie** [comment]
It will behind the custom delay-sum beamformer ?

**Comment by kamo-naoyuki** [comment]
I compared it with only beamformit, which is the baseline system for chime4, and got worse result than it.

**Comment by Rpersie** [comment]
I have run the baseline many times, but it is different. Sometimes, the results vary greatly.
![image](https://user-images.githubusercontent.com/12673020/59001294-932d1300-8840-11e9-9157-3c1fa9c03a22.png)
Do you think it may cause by this random?
If I just split the multichannel data to single channel and  dump them to json file, it will perform better and stably than the random training methods ?
Thank you

**Comment by sw005320** [comment]
> Do you think it may cause by this random?

Hmm.
We try to make things deterministic (specify a fixed seed for python, numpy, pytorch, chainer and use the deterministic mode).
We still observe some slight difference in the same machine and environment, but this difference seems to be too large.
@kamo-naoyuki, do you have comments?

**Comment by kamo-naoyuki** [comment]
I have no idea, sorry. All random seeds are fixed as I known, but  some bugs might be existing, I'm not sure because I didn't read all codes in espnet.

Could you show the loss or acc values among them? WER/CER is not the direct result, so I want to split the problem.

**Comment by Rpersie** [comment]
ok, below is the training log. The log label 1,2,3,4 crospond to the order of the table.
Additionly, the 2train.log is interrupted in 9epoch , and I resume training from 9epoch
[1train.log](https://github.com/espnet/espnet/files/3264537/1train.log)
[2train.log](https://github.com/espnet/espnet/files/3264539/2train.log)
[3train.log](https://github.com/espnet/espnet/files/3264540/3train.log)
[4train.log](https://github.com/espnet/espnet/files/3264541/4train.log)

**Comment by Rpersie** [comment]
@kamo-naoyuki @sw005320  can you give me some ideas how to solve or debug this problem ? I have some ideas to improve the baseline. But now I can not ensure if it is the effects of the improvement or just the random factors

**Comment by kamo-naoyuki** [comment]
> I suppose this is caused by the front-end random. But I cannot give strong evidence. I just split all the 5 channels into the training jsons to train the model. However, the dataset is too large and it will cost 4 days 1 gpu. In the original procedure, it just use the first channel .

The randomness at the frontend part is not relevant. The seed is  obviously fixed. You can easily check it without training. 
(It's not just using the first channel but using random channel at the feature transform part.)

I'll check the code tomorrow, just a moment.

**Comment by Rpersie** [comment]
I debug the program and find that in asr.py
`with open(args.train_json, 'rb') as f:
        train_json = json.load(f)['utts']
    with open(args.valid_json, 'rb') as f:
        valid_json = json.load(f)['utts']
`
I print the train_json and find it is different every training time.
if use ` train_json = json.load(f,object_pairs_hook=collections.OrderedDict)['utts']`.
So it is caused by this ? @kamo-naoyuki  . Thank you !

**Comment by kamo-naoyuki** [comment]
Yes. Maybe, you are right.

As of python3.7, the order of `dict` is deterministic.
As of Python3.5, or lower, the order is depending on the PYTHONHASHSEED. It means that the reulst of following code can be changed by the hash seed.
```python
print(list({'a': 0, 'b': 1, 'c': 2}))
```

For Python3.6, the behaviour may be different by its internal implementation.
As same as 3.5, the order is not deterministic as its strict specification, but actually, many python3.6 has same implementation as python3.7 as my experience. At least, my python3.6.8 can achieve same result in this codes.
Could you tell me your python version for my information? 

@sw005320 So, is it right that ESPNet should support python3.6+? Then, we must deal this problem as a bug.
We need to fix  at least two part, could you send a PR? @Rpersie (If you won't, I would)

- All json load part as you pointed. 
- Change the dict in the following part to OrderedDict

https://github.com/espnet/espnet/blob/6ee053f9935adf78f165c75386348c69dc643619/espnet/utils/training/batchfy.py#L323-L325

**Comment by Rpersie** [comment]
Ok, I will try to fix the bug, and test it before sending a PR? Thank you !
Expect this, I also find a bug in mask_estimator.py.
![image](https://user-images.githubusercontent.com/12673020/59170233-4d3dbb00-8b70-11e9-99b2-101886044a0c.png)
I think the Zero padding should be done after the sigmoid. Because in beamformer.py, you do the normalization like this:
![image](https://user-images.githubusercontent.com/12673020/59170331-a4dc2680-8b70-11e9-8c4d-2ebffe1b57e5.png)

**Comment by kamo-naoyuki** [comment]
As i noted there, it's no problem. The summation is same with zero padding.

**Comment by kamo-naoyuki** [comment]
Hmm, but why can you use Python3.5? These code doesn't work at python3.5 because fstring is used.
If you are really using Python3.5, you should Python3.6+. This is requirement of espnet.

**Comment by Rpersie** [comment]
no, sigmoid(0.0)  does not equal 0.So the mask should be done after sigmoid not before sigmoid . 
I just change the fstring

**Comment by kamo-naoyuki** [comment]
Oh, you are  right, sorry. sigmoid and padding should be swapped.
Use python3.6+. I can't guaranteed any results, so you don't need to fix the bug about reproducibility.

**Comment by Rpersie** [comment]
OK, I will try the python3.6+. Thank you !  Do you have ever try spliting the multichannel data into single  channels and dump into the json file. In this way , all five noisy channels can be used to train the acoustic model.

**Comment by kamo-naoyuki** [comment]
No I haven't, but all data can be used during many epochs because they are randomly picked.
Even if the recipe achieve better result than current, I think it's meaningless. It is accidental event.

**Comment by Rpersie** [comment]
ok! thank you! I want to do some contributions to this "multichannel ".  I want to implement the GEV beamformer, but the pytorch does not support svd.  Can you give some advices ? Or do you have some unfinished work i can help .

**Comment by kamo-naoyuki** [comment]
I was also interested in GEV (and I gave up.). so, maybe my advise is not so helpful. 
We need general eig decomposition for Complex Matrix, you know. But pytorch supports only

- (Not general) symmetric eigen value decomposition
- Real Tensor
https://pytorch.org/docs/stable/torch.html#torch.symeig

If not considering the speed, the most easy way may be using Tensorflow as this way: https://pytorch.org/tutorials/advanced/numpy_extensions_tutorial.html.
(I'm not familiar with tensorflow, so I don't know how to invoke the backward function actually.)

If you'll implement it in better pytorch way , then @ShigekiKarita's code is helpful: https://github.com/ShigekiKarita/pytorch-autograd-solver

**Comment by Rpersie** [comment]
So do you have any plan or unfinished ideas to improve this multichannel work ? eg. GAN, Teacher-Student Learning.

**Comment by kamo-naoyuki** [comment]
No, sorry. I'm tackling another tasks now. Actually, I'm a programmer rather than researcher.
If you are planning them, it could be great contribution for ESPNet. If you meet some issues at program level with ESPNet, I would help you.

Thank you for some reporting.

**Comment by Rpersie** [comment]
Ok, thanks

**Comment by sw005320** [comment]
Thanks a lot for your investigation.

> @sw005320 So, is it right that ESPNet should support python3.6+? Then, we must deal this problem as a bug.
We need to fix  at least two part, could you send a PR? @Rpersie (If you won't, I would)
>
> - All json load part as you pointed. 
> - Change the dict in the following part to OrderedDict
> 
> https://github.com/espnet/espnet/blob/6ee053f9935adf78f165c75386348c69dc643619/espnet/utils/training/batchfy.py#L323-L325

Yes, it is right.
Could either of you make a PR about this?

**Comment by kamo-naoyuki** [comment]
I'm rethinking whether it should be fixed... because it may not happen on CPython3.6. 
OrderedDict usage looks a little ugly, and we need to use `dict` very carefully since now if we deal this problem as a bug even though it's really nothing.

I'll send PR about the padding bug after checking the performance changing.

**Comment by Rpersie** [comment]
As the machine in our lab is not collected with the internet, it is troublesome to install the software offline.
So I try to fix this dict on my version and test the results. I check  the input data, now it is the same. 
While I run two experiments at the same, the loss is different. I think it is caused by the pytorch initiation ? Am I right ?  @kamo-naoyuki 
I will wait the final results.  If you want a PR, then I will pull. But I need to test it.  thank you !

**Comment by Rpersie** [comment]
Sorry, I use different machine. I will test it more.

**Comment by kamo-naoyuki** [comment]
> OK, I will try the python3.6+. Thank you ! Do you have ever try spliting the multichannel data into single channels and dump into the json file. In this way , all five noisy channels can be used to train the acoustic model.

@Rpersie Have your training with the splitting data been fished? I'll also try it due to another reason, so I'd like you to tell me your result.

**Comment by Rpersie** [comment]
Yeah, but the results are not very well. 
![image](https://user-images.githubusercontent.com/12673020/59986440-73cc1d80-9669-11e9-8706-b479bc08a419.png)

**Comment by Rpersie** [comment]
If you need it, I will try just split data and do not change the front-end code.

**Comment by kamo-naoyuki** [comment]
Thanks! The degradation looks mysterious.

**Comment by Rpersie** [comment]
@kamo-naoyuki  Recently, I implement the MVDR using steervector , but the espnet have the warnings in every epoch "grad norm is nan. Do not update model".  have you ever countered this problem ? Thank you !

**Comment by Rpersie** [comment]
I just check you code and plot the enhanced FFT Spectrogram. I just exit(1) the program after the beamforming. So the figure is just using the initialized parameters and the grad is not updated.
The first figure is the espnet results and the second is my improved results. I find my results is better 
than the espnet. But it  has the warnings in every epoch "grad norm is nan. Do not update model".
Can you give me some advices how to debug it ? Thank you. 
@kamo-naoyuki  @sw005320 

![image](https://user-images.githubusercontent.com/12673020/61205085-afff1500-a721-11e9-9255-f043e51b4480.png)
![image](https://user-images.githubusercontent.com/12673020/61205109-c0af8b00-a721-11e9-9543-3766cc3fc16d.png)

**Comment by sw005320** [comment]
> @kamo-naoyuki Recently, I implement the MVDR using steervector , but the espnet have the warnings in every epoch "grad norm is nan. Do not update model". have you ever countered this problem ? Thank you !

This might be a bug due to your implementation. How did you get a steering vector? I believe we did not fully support this.
